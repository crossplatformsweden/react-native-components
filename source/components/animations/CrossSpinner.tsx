import React from 'react';
import {
  BallIndicator,
  BarIndicator,
  DotIndicator,
  MaterialIndicator,
  PacmanIndicator,
  PulseIndicator,
  SkypeIndicator,
  UIActivityIndicator,
  WaveIndicator,
  BallIndicatorProps,
} from 'react-native-indicators';

import { Colors } from '../../styles';

/**
 * Type of spinner. See examples at:
 *
 * https://github.com/n4kz/react-native-indicators
 *
 * @export
 * @enum {string}
 */
export enum CrossSpinnerType {
  BallIndicator = 'BallIndicator',
  BarIndicator = 'BarIndicator',
  DotIndicator = 'DotIndicator',
  MaterialIndicator = 'MaterialIndicator',
  PacmanIndicator = 'PacmanIndicator',
  PulseIndicator = 'PulseIndicator',
  SkypeIndicator = 'SkypeIndicator',
  UIActivityIndicator = 'UIActivityIndicator',
  WaveIndicator = 'WaveIndicator',
}

export interface ICrossSpinnerProps extends BallIndicatorProps {
  /**
   * {@link IndicatorType}
   */
  type: CrossSpinnerType;
  /**
   * Optional custom styles
   */
  style?: any;
  testID?: string|undefined;
}

/**
 * Generates a React-Native spinning busy indicator
 *
 * https://github.com/n4kz/react-native-indicators
 *
 * @param param0 see {@link ISpinnerProps}
 */
export const CrossSpinner = ({
  type,
  style = null,
  color = Colors.CrossLightBlue,
  count = 8,
  size = 40,
  testID = type.toString()
}: ICrossSpinnerProps) => {
  const props: ICrossSpinnerProps = {
    type,
    style,
    count,
    size,
    color,
    testID
  };
  switch (type) {
    case 'BallIndicator':
      return <BallIndicator {...props} />;
    case 'BarIndicator':
      return <BarIndicator {...props} />;
    case 'DotIndicator':
      return <DotIndicator {...props} />;
    case 'PacmanIndicator':
      return <PacmanIndicator {...props} />;
    case 'PulseIndicator':
      return <PulseIndicator {...props} />;
    case 'SkypeIndicator':
      return <SkypeIndicator {...props} />;
    case 'UIActivityIndicator':
      return <UIActivityIndicator {...props} />;
    case 'WaveIndicator':
      return <WaveIndicator {...props} />;
    default:
      return <MaterialIndicator {...props} />;
  }
};
