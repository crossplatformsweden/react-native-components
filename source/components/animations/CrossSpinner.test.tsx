import React from 'react';
import TestRenderer from 'react-test-renderer';
import { CrossSpinner, CrossSpinnerType } from './CrossSpinner';
import {
  BallIndicator,
  BarIndicator,
  DotIndicator,
  MaterialIndicator,
  PacmanIndicator,
  PulseIndicator,
  SkypeIndicator,
  UIActivityIndicator,
  WaveIndicator,
} from 'react-native-indicators';

function setup(type?: CrossSpinnerType, size?: number, color?: string) {
  return TestRenderer.create(
    <CrossSpinner
      type={type || 'BallIndicator'}
      color={color || 'blue'}
      size={size || 64}
    />
  );
}

jest.unmock('react-native');
jest.unmock('./CrossSpinner');

describe('components', () => {
  describe('<CrossSpinner />', () => {
    describe('`CrossSpinnerType` enum', () => {
      it('Should have `BallIndicator` enum with value "BallIndicator"', () => {
        expect('BallIndicator').toEqual('BallIndicator');
      });
    });

    it('Component should render', () => {
      const wrapper = setup();
      expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('Should have <BallIndicator /> with size 64', () => {
      const wrapper = setup('BallIndicator');
      const child = wrapper.root.findByType(BallIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <BarIndicator /> with size 64', () => {
      const wrapper = setup();
      const child = wrapper.root.findByType(BarIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <DotIndicator /> with size 64', () => {
      const wrapper = setup('DotIndicator');
      const child = wrapper.root.findByType(DotIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <PacmanIndicator /> with size 64', () => {
      const wrapper = setup('PacmanIndicator');
      const child = wrapper.root.findByType(PacmanIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <PulseIndicator /> with size 64', () => {
      const wrapper = setup('PulseIndicator');
      const child = wrapper.root.findByType(PulseIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <SkypeIndicator /> with size 64', () => {
      const wrapper = setup('SkypeIndicator');
      const child = wrapper.root.findByType(SkypeIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <UIActivityIndicator /> with size 64', () => {
      const wrapper = setup('UIActivityIndicator');
      const child = wrapper.root.findByType(UIActivityIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <WaveIndicator /> with size 64', () => {
      const wrapper = setup('WaveIndicator');
      const child = wrapper.root.findByType(WaveIndicator);
      expect(child.props.size).toEqual(64);
    });

    it('Should have <MaterialIndicator /> with size 64', () => {
      const wrapper = setup('MaterialIndicator');
      const child = wrapper.root.findByType(MaterialIndicator);
      expect(child.props.size).toEqual(64);
    });
  });
});
