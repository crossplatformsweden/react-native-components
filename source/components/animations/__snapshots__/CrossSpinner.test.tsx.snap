// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <CrossSpinner /> \`CrossSpinnerType\` enum Should match snapshot 1`] = `
Object {
  "BallIndicator": "BallIndicator",
  "BarIndicator": "BarIndicator",
  "DotIndicator": "DotIndicator",
  "MaterialIndicator": "MaterialIndicator",
  "PacmanIndicator": "PacmanIndicator",
  "PulseIndicator": "PulseIndicator",
  "SkypeIndicator": "SkypeIndicator",
  "UIActivityIndicator": "UIActivityIndicator",
  "WaveIndicator": "WaveIndicator",
}
`;

exports[`components <CrossSpinner /> Component should render 1`] = `
<View
  color="blue"
  count={8}
  size={64}
  style={null}
  testID="BallIndicator"
  type="BallIndicator"
/>
`;
